k8sclient/k8sclient.go

```diff
@@ -511,24 +504,31 @@ func GetPodNetwork(k8sclient KubeClient, k8sArgs *types.K8sArgs, confdir string,
 	if err != nil {
 		return nil, err
 	}
+	return networks, nil
+}
 
+func GetNetworkDelegates(k8sclient KubeClient, pod *v1.Pod, networks []*types.NetworkSelectionElement, confdir string, confnamespaceIsolation bool) ([]*types.DelegateNetConf, error) {
+	logging.Debugf("GetNetworkDelegates: %v, %v, %v, %v, %v", k8sclient, pod, networks, confdir, confnamespaceIsolation)
 	// resourceMap holds Pod device allocation information; only initizized if CRD contains 'resourceName' annotation.
 	// This will only be initialized once and all delegate objects can reference this to look up device info.
 	var resourceMap map[string]*types.ResourceInfo
 
 	// Read all network objects referenced by 'networks'
 	var delegates []*types.DelegateNetConf
+	defaultNamespace := pod.ObjectMeta.Namespace
+
+	podID := pod.UID
```

Comment:
Instead of using the pod.UID to track caching, I would use the sandbox ID that is passed via the CNI args as the "container ID". The problem with using the Pod UID is that it will be the same for all sandboxes for the pod, and you can have overlapping sandboxes if one fails. Kubelet garbage-collects dead sandboxes asynchronously.

So for example, you start a pod and there's a transient error creating the sandbox network. Kubelet will schedule that one for garbage collection but not call CNI DEL yet. Then after a short period of time it will create another sandbox and call CNI ADD and perhaps this one works. Now the original one gets GC-ed but the second is still alive.

So if you use the Pod UID as a key, you might delete data you had cached that still applies to the second sandbox. But each sandbox will have a unique container ID (CNI's CNI_CONTAINERID) that you can use to ensure stuff like this doesn't overlap.